Spurious Warning when Using -pedantic-errors / -Wnewline-eof with Clang

### Xmake Version

xmake v2.9.7+20250112

### Operating System Version and Architecture

Fedora 41 x64

### Describe Bug

Clang has an option named -pedantic-errors that throws errors (as opposed to warnings; cf. -pedantic) when the source code deviates from the C/C++ standard. One diagnostic that is turned on when using -pedantic-errors is -Wnewline-eof, which is shown when the file doesn't end in a newline.

xmake tries to compile a sample file to check most flags (from _check_try_running). The content is `"int main(int argc, char** argv)\n{return 0;}"`. As the file does not end in a new line, it does not compile when checking whether -pedantic-errors is a valid flag (or -Wnewline-eof for that matter) because the flag correctly throws an error but the code misinterprets is as an invalid flag.

Changing the snippet file to add a new line at the end fixes this issue.

Note: Only clang shows this behaviour, as gcc does not have a -Wnewline-eof option.

### Expected Behavior

No warning when setting -pedantic-errors.

### Project Configuration

```lua
target("test")
    set_kind("binary")
    add_files("*.c")
    set_toolchains("clang")
    add_cflags("-pedantic-errors")
```

### Additional Information and Error Logs

```console
checking for clang ... /usr/lib64/ccache/clang
checking for the c compiler (cc) ... clang
checking for /usr/lib64/ccache/clang ... ok
checking for flags (-fPIC) ... ok
> clang "-fPIC" "-Qunused-arguments" "-m64"
checking for flags (-pedantic-errors) ... no
> clang "-pedantic-errors" "-Qunused-arguments" "-m64"
checkinfo: @programdir/core/sandbox/modules/os.lua:273: /tmp/.xmake1000/250112/_F3443E45635A466AA3BEAE9DE99B4339.c:2:12: error: no newline at end of file [-Werror,-Wnewline-eof]
    2 | {return 0;}
      |            ^
1 error generated.

stack traceback:
    [C]: in function 'error'
    [@programdir/core/base/os.lua:1004]:
    [@programdir/core/sandbox/modules/os.lua:273]: in function 'runv'
    [@programdir/modules/detect/tools/gcc/has_flags.lua:42]:
create ok!
warning: add_cflags("-pedantic-errors") is ignored, please pass `{force = true}` or call `set_policy("check.auto_ignore_flags", false)` if you want to set it.
```

--- a/xmake/modules/detect/tools/armclang/has_flags.lua
+++ b/xmake/modules/detect/tools/armclang/has_flags.lua
@@ -90,7 +90,7 @@ function _check_try_running(flags, opt, islinker)
     -- make an stub source file
     local sourcefile = path.join(os.tmpdir(), "detect", "armclang_has_flags" .. _get_extension(opt))
     if not os.isfile(sourcefile) then
-        io.writefile(sourcefile, "int main(int argc, char** argv)\n{return 0;}")
+        io.writefile(sourcefile, "int main(int argc, char** argv)\n{return 0;}\n")
     end
 
     -- check flags for linker

--- a/xmake/modules/detect/tools/cl/has_flags.lua
+++ b/xmake/modules/detect/tools/cl/has_flags.lua
@@ -64,7 +64,7 @@ end
 function _check_try_running(flags, opt)
 
     -- make an stub source file
-    local snippet = opt.snippet or "int main(int argc, char** argv)\n{return 0;}"
+    local snippet = opt.snippet or "int main(int argc, char** argv)\n{return 0;}\n"
     local sourcefile = os.tmpfile("cl_has_flags:" .. snippet) .. _get_extension(opt)
     if not os.isfile(sourcefile) then
         io.writefile(sourcefile, snippet)

--- a/xmake/modules/detect/tools/clang_cl/has_flags.lua
+++ b/xmake/modules/detect/tools/clang_cl/has_flags.lua
@@ -74,7 +74,7 @@ function _check_try_running(flags, opt)
     -- make an stub source file
     local sourcefile = path.join(os.tmpdir(), "detect", "clang_cl_has_flags" .. extension)
     if not os.isfile(sourcefile) then
-        io.writefile(sourcefile, "int main(int argc, char** argv)\n{return 0;}")
+        io.writefile(sourcefile, "int main(int argc, char** argv)\n{return 0;}\n")
     end
 
     -- check flags for compiler

--- a/xmake/modules/detect/tools/cosmocc/has_flags.lua
+++ b/xmake/modules/detect/tools/cosmocc/has_flags.lua
@@ -92,7 +92,7 @@ end
 function _check_try_running(flags, opt, islinker)
 
     -- make an stub source file
-    local snippet = opt.snippet or "int main(int argc, char** argv)\n{return 0;}"
+    local snippet = opt.snippet or "int main(int argc, char** argv)\n{return 0;}\n"
     local sourcefile = os.tmpfile("cosmocc_has_flags:" .. snippet) .. _get_extension(opt)
     if not os.isfile(sourcefile) then
         io.writefile(sourcefile, snippet)

--- a/xmake/modules/detect/tools/find_cl.lua
+++ b/xmake/modules/detect/tools/find_cl.lua
@@ -44,7 +44,7 @@ function main(opt)
             -- @see https://github.com/xmake-io/xmake/issues/3057
             local objectfile = os.tmpfile() .. ".obj"
             local sourcefile = os.tmpfile() .. ".c"
-            io.writefile(sourcefile, "int main(int argc, char** argv)\n{return 0;}")
+            io.writefile(sourcefile, "int main(int argc, char** argv)\n{return 0;}\n")
             os.runv(program, {"-c", "-Fo" .. objectfile, sourcefile}, {envs = opt.envs})
             os.rm(objectfile)
             os.rm(sourcefile)

--- a/xmake/modules/detect/tools/find_emar.lua
+++ b/xmake/modules/detect/tools/find_emar.lua
@@ -60,7 +60,7 @@ function main(opt)
         local libraryfile   = os.tmpfile() .. ".a"
         local objectfile    = os.tmpfile() .. ".o"
         local sourcefile    = os.tmpfile() .. ".c"
-        io.writefile(sourcefile, "int test(void)\n{return 0;}")
+        io.writefile(sourcefile, "int test(void)\n{return 0;}\n")
 
         -- compile it
         os.runv(emcc.program, {"-c", "-o" .. objectfile, sourcefile}, {envs = opt.envs})

--- a/xmake/modules/detect/tools/find_lib.lua
+++ b/xmake/modules/detect/tools/find_lib.lua
@@ -48,7 +48,7 @@ function main(opt)
         local libraryfile = os.tmpfile() .. ".lib"
         local objectfile  = os.tmpfile() .. ".obj"
         local sourcefile  = os.tmpfile() .. ".c"
-        io.writefile(sourcefile, "int test(void)\n{return 0;}")
+        io.writefile(sourcefile, "int test(void)\n{return 0;}\n")
 
         -- check it
         local cl = assert(find_tool("cl", {envs = opt.envs}))

--- a/xmake/modules/detect/tools/find_link.lua
+++ b/xmake/modules/detect/tools/find_link.lua
@@ -60,7 +60,7 @@ function main(opt)
             local sourcefile = os.tmpfile() .. ".c"
 
             -- compile sourcefile first
-            io.writefile(sourcefile, "int main(int argc, char** argv)\n{return 0;}")
+            io.writefile(sourcefile, "int main(int argc, char** argv)\n{return 0;}\n")
             os.runv(cl.program, {"-c", "-Fo" .. objectfile, sourcefile}, {envs = opt.envs})
 
             -- do link

--- a/xmake/modules/detect/tools/find_strip.lua
+++ b/xmake/modules/detect/tools/find_strip.lua
@@ -28,7 +28,7 @@ function _check_strip_of_xcode(program)
     -- make an stub source file
     local objectfile    = os.tmpfile() .. ".o"
     local sourcefile    = os.tmpfile() .. ".c"
-    io.writefile(sourcefile, "int test(void)\n{return 0;}")
+    io.writefile(sourcefile, "int test(void)\n{return 0;}\n")
 
     -- compile it
     compiler.compile(sourcefile, objectfile)

--- a/xmake/modules/detect/tools/gcc/has_flags.lua
+++ b/xmake/modules/detect/tools/gcc/has_flags.lua
@@ -89,7 +89,7 @@ end
 function _check_try_running(flags, opt, islinker)
 
     -- make an stub source file
-    local snippet = opt.snippet or "int main(int argc, char** argv)\n{return 0;}"
+    local snippet = opt.snippet or "int main(int argc, char** argv)\n{return 0;}\n"
     local sourcefile = os.tmpfile("gcc_has_flags:" .. snippet) .. _get_extension(opt)
     if not os.isfile(sourcefile) then
         io.writefile(sourcefile, snippet)

--- a/xmake/modules/detect/tools/link/has_flags.lua
+++ b/xmake/modules/detect/tools/link/has_flags.lua
@@ -63,9 +63,9 @@ function _check_try_running(flags, opt)
     local sourcefile = path.join(os.tmpdir(), "detect", (winmain and "winmain_" or "") .. "link_has_flags.c")
     if not os.isfile(sourcefile) then
         if winmain then
-            io.writefile(sourcefile, "int WinMain(void* instance, void* previnst, char** argv, int argc)\n{return 0;}")
+            io.writefile(sourcefile, "int WinMain(void* instance, void* previnst, char** argv, int argc)\n{return 0;}\n")
         else
-            io.writefile(sourcefile, "int main(int argc, char** argv)\n{return 0;}")
+            io.writefile(sourcefile, "int main(int argc, char** argv)\n{return 0;}\n")
         end
     end
 

--- a/xmake/modules/detect/tools/nvcc/has_flags.lua
+++ b/xmake/modules/detect/tools/nvcc/has_flags.lua
@@ -101,7 +101,7 @@ end
 function _check_try_running(flags, opt, islinker)
 
     -- make an stub source file
-    local snippet = opt.snippet or "int main(int argc, char** argv)\n{return 0;}"
+    local snippet = opt.snippet or "int main(int argc, char** argv)\n{return 0;}\n"
     local sourcefile = os.tmpfile("nvcc_has_flags:" .. snippet) .. ".cu"
     if not os.isfile(sourcefile) then
         io.writefile(sourcefile, snippet)

--- a/xmake/modules/detect/tools/sdcc/has_flags.lua
+++ b/xmake/modules/detect/tools/sdcc/has_flags.lua
@@ -88,7 +88,7 @@ function _check_try_running(flags, opt, islinker)
     -- make an stub source file
     local sourcefile = path.join(os.tmpdir(), "detect", "sdcc_has_flags" .. extension)
     if not os.isfile(sourcefile) then
-        io.writefile(sourcefile, "int main(int argc, char** argv)\n{return 0;}")
+        io.writefile(sourcefile, "int main(int argc, char** argv)\n{return 0;}\n")
     end
 
     -- check flags for linker

